### PR TITLE
1607 metric supporting autoscaling

### LINKED ISSUE #1607: Expose metric that can be used to autoscale workers
Expose a metric that allows autoscaling when the workers take strain.

### PR DESCRIPTION
**Pull request by rorymckinley**
## Notes for the reviewer
This branch contains some cleanup/rearranging work both before the actual change and afterwards. This makes the diffs a little bit noisier than they should be. In light of this, I tried to have each commit only cover a single type of change (i.e. cleanup or new functionality) and my recommendation would be to review commit-by-commit to ease understanding.

I will collapse the commits once the PR has been approved.

## Related issue

Fixes #1607

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

### CODE REVIEW

File: lib/lightning/attempts/prom_ex_plugin.ex [hunk 1]
```diff
@@ -78,4 +83,89 @@ defmodule Lightning.Attempts.PromExPlugin do
...
+  defp claim_duration(attempt, _reference_time) do
+    DateTime.diff(attempt.claimed_at, attempt.inserted_at, :millisecond)
+  end
+
+  defp average({_sum, 0}) do
+    0.0
+  end
+
+  defp average({sum, count}) do
+    (sum / count) |> Float.round(0)
```
Comment (jyeshe) [inline lib/lightning/attempts/prom_ex_plugin.ex:161]: as the unit is `millisecond` if we can would use integer format to replace the decimal with 0 digits: `trunc(sum / count)`

File: lib/lightning/prom_ex.ex [hunk 2]
```diff
@@ -58,6 +58,16 @@ defmodule Lightning.PromEx do
...
   def plugins do
+    stalled_attempt_threshold_seconds =
+      Application.get_env(:lightning, :metrics)[
+        :stalled_attempt_threshold_seconds
+      ]
+
+    attempt_performance_age_seconds =
+      Application.get_env(:lightning, :metrics)[
+        :attempt_performance_age_seconds
+      ]
```
Comment (jyeshe) [inline lib/lightning/prom_ex.ex:69]: would give smaller names here too.

**Review by taylordowns2000 — APPROVED** [review]

File: lib/lightning/attempts/prom_ex_plugin.ex [hunk 3]
```diff
@@ -78,4 +83,89 @@ defmodule Lightning.Attempts.PromExPlugin do
...
+
+  defp trigger_attempt_claim_duration(nil, _attempt_age_seconds) do
+    nil
+  end
+
+  defp trigger_attempt_claim_duration(repo_pid, attempt_age_seconds) do
+    check_repo_state(repo_pid)
+
+    average_duration =
+      calculate_average_claim_duration(DateTime.utc_now(), attempt_age_seconds)
```
Comment (jyeshe) [inline lib/lightning/attempts/prom_ex_plugin.ex:123]: as it's always now would remove the param.

File: lib/lightning/attempts/prom_ex_plugin.ex [hunk 4]
```diff
@@ -78,4 +83,89 @@ defmodule Lightning.Attempts.PromExPlugin do
...
+    |> average()
+  end
+
+  defp claim_duration(%Attempt{state: :available} = attempt, reference_time) do
+    DateTime.diff(reference_time, attempt.inserted_at, :millisecond)
+  end
+
+  defp claim_duration(attempt, _reference_time) do
+    DateTime.diff(attempt.claimed_at, attempt.inserted_at, :millisecond)
+  end
```
Comment (jyeshe) [inline lib/lightning/attempts/prom_ex_plugin.ex:154]: :+1:

**Review by jyeshe — COMMENTED** [review]
Don't have the environment to test but looks solid and well tested. Suggestions on readability only.

**Pull request merged**